File: wrench-app/src/main/java/com/izettle/wrench/dialogs/enumvalue/FragmentEnumValueViewModel.java
package com.izettle.wrench.dialogs.enumvalue;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;

import com.izettle.wrench.database.WrenchConfiguration;
import com.izettle.wrench.database.WrenchConfigurationDao;
import com.izettle.wrench.database.WrenchConfigurationValue;
import com.izettle.wrench.database.WrenchConfigurationValueDao;
import com.izettle.wrench.database.WrenchPredefinedConfigurationValue;
import com.izettle.wrench.database.WrenchPredefinedConfigurationValueDao;

import java.util.Date;
import java.util.List;

import javax.inject.Inject;

public class FragmentEnumValueViewModel extends ViewModel {

    private final WrenchConfigurationDao configurationDao;
    private final WrenchConfigurationValueDao configurationValueDao;
    private WrenchPredefinedConfigurationValueDao predefinedConfigurationValueDao;
    private LiveData<WrenchConfiguration> configuration;
    private long configurationId;
    private long scopeId;
    private LiveData<WrenchConfigurationValue> selectedConfigurationValueLiveData;
    private WrenchConfigurationValue selectedConfigurationValue;
    private LiveData<List<WrenchPredefinedConfigurationValue>> predefinedValuesLiveData;

    @SuppressWarnings("WeakerAccess")
    @Inject
    public FragmentEnumValueViewModel(WrenchConfigurationDao configurationDao, WrenchConfigurationValueDao configurationValueDao, WrenchPredefinedConfigurationValueDao predefinedConfigurationValueDao) {
        this.configurationDao = configurationDao;
        this.configurationValueDao = configurationValueDao;
        this.predefinedConfigurationValueDao = predefinedConfigurationValueDao;
    }

    void init(long configurationId, long scopeId) {
        this.configurationId = configurationId;
        this.scopeId = scopeId;
    }

    LiveData<WrenchConfiguration> getConfiguration() {
        if (configuration == null) {
            configuration = configurationDao.getConfiguration(configurationId);
        }
        return configuration;
    }

    public void updateConfigurationValue(String value) {
        if (selectedConfigurationValue != null) {
            configurationValueDao.updateConfigurationValue(configurationId, scopeId, value);
        } else {
            WrenchConfigurationValue wrenchConfigurationValue = new WrenchConfigurationValue(0, configurationId, value, scopeId);
            wrenchConfigurationValue.setId(configurationValueDao.insert(wrenchConfigurationValue));
        }
        configurationDao.touch(configurationId, new Date());
    }

    void deleteConfigurationValue() {
        configurationValueDao.delete(selectedConfigurationValue);
    }

    LiveData<WrenchConfigurationValue> getSelectedConfigurationValueLiveData() {
        if (selectedConfigurationValueLiveData == null) {
            selectedConfigurationValueLiveData = configurationValueDao.getConfigurationValue(configurationId, scopeId);
        }
        return selectedConfigurationValueLiveData;
    }

    LiveData<List<WrenchPredefinedConfigurationValue>> getPredefinedValues() {
        if (predefinedValuesLiveData == null) {
            predefinedValuesLiveData = predefinedConfigurationValueDao.getByConfigurationId(configurationId);
        }
        return predefinedValuesLiveData;
    }

    WrenchConfigurationValue getSelectedConfigurationValue() {
        return selectedConfigurationValue;
    }

    void setSelectedConfigurationValue(WrenchConfigurationValue selectedConfigurationValue) {
        this.selectedConfigurationValue = selectedConfigurationValue;
    }
}

File: wrench-app/src/main/java/com/izettle/wrench/dialogs/enumvalue/FragmentEnumValueViewModel.kt
package com.izettle.wrench.dialogs.enumvalue

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import com.izettle.wrench.database.*
import java.util.*
import javax.inject.Inject

class FragmentEnumValueViewModel @Inject
constructor(
        private val configurationDao: WrenchConfigurationDao,
        private val configurationValueDao: WrenchConfigurationValueDao,
        private val predefinedConfigurationValueDao: WrenchPredefinedConfigurationValueDao)
    : ViewModel() {

    internal val configuration: LiveData<WrenchConfiguration> by lazy {
        configurationDao.getConfiguration(configurationId)
    }

    internal val selectedConfigurationValueLiveData: LiveData<WrenchConfigurationValue>  by lazy {
        configurationValueDao.getConfigurationValue(configurationId, scopeId)
    }

    internal val predefinedValues: LiveData<List<WrenchPredefinedConfigurationValue>> by lazy {
        predefinedConfigurationValueDao.getByConfigurationId(configurationId)
    }

    private var configurationId: Long = 0
    private var scopeId: Long = 0

    internal var selectedConfigurationValue: WrenchConfigurationValue? = null

    internal fun init(configurationId: Long, scopeId: Long) {
        this.configurationId = configurationId
        this.scopeId = scopeId
    }

    fun updateConfigurationValue(value: String) {
        if (selectedConfigurationValue != null) {
            configurationValueDao.updateConfigurationValue(configurationId, scopeId, value)
        } else {
            val wrenchConfigurationValue = WrenchConfigurationValue(0, configurationId, value, scopeId)
            wrenchConfigurationValue.id = configurationValueDao.insert(wrenchConfigurationValue)
        }
        configurationDao.touch(configurationId, Date())
    }

    internal fun deleteConfigurationValue() {
        configurationValueDao.delete(selectedConfigurationValue!!)
    }
}
